src/useCustomEffect.ts

```diff
@@ -0,0 +1,91 @@
+import {
+  DependencyList,
+  EffectCallback,
+  useRef,
+  useEffect,
+  useDebugValue,
+} from 'react'
+import useWillUnmount from './useWillUnmount'
+import useMounted from './useMounted'
+
+export type EffectHook = (effect: EffectCallback, deps?: DependencyList) => void
+
+export type IsEqual<TDeps> = (nextDeps: TDeps, prev: TDeps) => boolean
+
+export type CustomEffectOptions<TDeps> = {
+  isEqual: IsEqual<TDeps>
+  effectHook?: EffectHook
+}
+
+type CleanUp = {
+  (): void
+  cleanup?: ReturnType<EffectCallback>
+}
+
+/**
+ * a useEffect() hook with customized depedency comparision
+ *
+ * @param effect The effect callback
+ * @param dependencies A list of dependencies
+ * @param isEqual A function comparing the next and previous dependencyLists
+ */
+function useCustomEffect<TDeps extends DependencyList = DependencyList>(
```

Comment:
this typing seems odd to me. `DependencyList` is typed as `ReadonlyArray<any>`... but realistically we're going to have something like a tuple type. do tuple types extend read-only arrays of type any?

... apparently so, but it's weird